imports/importers.py

```diff
@@ -48,49 +67,116 @@ class ShapefileImporter:
+
+    def _validate_fields(self, layer):
+        """
+        Validate layer fields
+
+        :param layer: The layer to be validated
+        :raise ImportValidationError: if invalid field found
+        """
+        layer_fields = layer.fields[1:]  # first field reserved as deletion flag
+        fields_not_allowed = set(layer_fields) - set(SHAPEFILE_FIELD_MAPPING.keys())
+        if fields_not_allowed:
+            error_code = "fields_not_allowed"
+            raise ImportValidationError(
+                self.error_messages[error_code],
+                code=error_code,
+                params={
+                    "fields": ", ".join(fields_not_allowed),
+                    "allowed_fields": ", ".join(SHAPEFILE_FIELD_MAPPING.keys()),
+                },
             )
-            with zfile.open(dbf) as dbf, zfile.open(shp) as shp, zfile.open(shx) as shx:
-                return shapefile.Reader(shp=shp, shx=shx, dbf=dbf)
 
-    @classmethod
-    def _import_feature(cls, fields, shape_record):
-        feature_data = dict(zip(fields, shape_record.record))
-        feature_data["geometry"] = cls._get_feature_geometry(shape_record.shape)
-
-        feature_id = feature_data.pop("id", None)
-        try:
-            feature = Feature.objects.get(id=feature_id)
-            for key, value in feature_data.items():
-                setattr(feature, key, value)
-                feature.save()
-        except Feature.DoesNotExist:
-            Feature.objects.create(**feature_data)
+    def _validate_feature_classes(self, layer):
+        """
+        Validate layer feature classes
 
-    @classmethod
-    def _get_feature_geometry(cls, shape):
-        shape_type = shape.__geo_interface__["type"].upper()
-        coordinates = "{}".format(shape.__geo_interface__["coordinates"]).replace(
-            ",", " "
+        :param layer: The layer to be validated
+        :raise ImportValidationError: if invalid feature class found
+        """
+        layer_feature_classes = set(layer.get_fields("luokkatunn"))
+        db_feature_classes = set(
+            FeatureClass.objects.filter(id__in=layer_feature_classes).values_list(
+                "id", flat=True
+            )
         )
-        geostr = "SRID={};{}{}".format(settings.SRID, shape_type, coordinates)
-        return GEOSGeometry(geostr, srid=settings.SRID)
+        missing_feature_classes = layer_feature_classes - db_feature_classes
+        if missing_feature_classes:
+            error_code = "feature_classes_not_exist"
+            raise ImportValidationError(
+                self.error_messages[error_code],
+                code=error_code,
+                params={"feature_classes": ", ".join(missing_feature_classes)},
+            )
+
+    def _import_layer(self, layer):
+        """
+        Import features from shapefiles layer
+
+        :param shapefiles: The path to the shapefiles shp file
+        :return: The number of features imported
+        """
+        count = 0
```

Comment:
I don't see any `continue` or `break` in the for loop below, nor is the count updated in the inner loop. Won't `count` just always be `len(layer)` here?

Reply:
That's a good point :+1:

Reply:
That's a good point +1